File: compiler/Sources/cppc/compiler/utils/ConfigurationManager.java
//    This file is part of CPPC.
//
//    CPPC is free software; you can redistribute it and/or modify
//    it under the terms of the GNU General Public License as published by
//    the Free Software Foundation; either version 2 of the License, or
//    (at your option) any later version.
//
//    CPPC is distributed in the hope that it will be useful,
//    but WITHOUT ANY WARRANTY; without even the implied warranty of
//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//    GNU General Public License for more details.
//
//    You should have received a copy of the GNU General Public License
//    along with CPPC; if not, write to the Free Software
//    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA




package cppc.compiler.utils;

import java.util.HashMap;

public final class ConfigurationManager {

  private static HashMap<String,String[]> options =
    new HashMap<String,String[]>();
  private static final String CPPC_ARG_PREFIX = "-CPPC,";

  // Options
  public static final String DISABLE_CKPT_ANALYSIS = "DisableCkptAnalysis";
  public static final String DISABLE_COMM_ANALYSIS = "DisableCommAnalysis";
  public static final String MANUAL_PRAGMAS_OPTION = "ManualPragmas";
  public static final String PROCESS_NUMBER_OPTION = "ProcessNumber";

  private ConfigurationManager() {}

  public static final String[] parseCommandLine( String[] args ) {

    if( args.length == 0 ) {
      return args;
    }

    for( int i = 0; i < args.length; i++ ) {

      String option = args[i].trim();

      // Is this our option?
      if( !option.startsWith( CPPC_ARG_PREFIX ) ) {
        continue;
      }

      // If it is, remove the header
      option = option.replaceFirst( CPPC_ARG_PREFIX, "" );
      option.trim();

      // Split the remaining parts
      String[] subOptions = option.split( "," );

      for( int j = 0; j < subOptions.length; j++ ) {

        String subOption = subOptions[j].trim();

        String[] expr = subOption.split( "=" );

        if( expr.length != 2 ) {
          System.err.println( "WARNING: Ignoring command-line argument '" +
            subOption + "' : Syntax error" );
        } else {
          String key = expr[0].trim();
          String value = expr[1].trim();
          if( options.containsKey( key ) ) {
            String [] values = options.get( key );
            String [] newValues= new String[ values.length+1 ];
            System.arraycopy( values, 0, newValues, 0, values.length );
            newValues[ values.length ] = value;
            options.put( key, newValues );
          } else {
            String [] values = { value };
            options.put( key, values );
          }
        }
      }

      // Remove this option from the list
      String[] newArgs = new String[ args.length - 1 ];
      for( int j = 0; j < i; j++ ) {
        newArgs[ j ] = args[ j ];
      }
      for( int j = i; j < args.length-1; j++ ) {
        newArgs[j] = args[j+1];
      }

      return newArgs;
    }

    return args;
  }

  public static final boolean hasOption( String key ) {
    return options.containsKey( key );
  }

  public static final String getOption( String key ) {
    if( options.containsKey( key ) ) {
      return options.get( key )[ 0 ];
    } else {
      return null;
    }
  }

  public static final String[] getOptionArray( String key ) {
    if( options.containsKey( key ) ) {
      return options.get( key );
    } else {
      return new String[0];
    }
  }

  public static final void setOption( String key, String value ) {
    String [] values = { value.trim() };
    options.put( key.trim(), values );
  }
}

File: src/util/configuration/configuration_manager.cpp
//    This file is part of CPPC.
//
//    CPPC is free software; you can redistribute it and/or modify
//    it under the terms of the GNU General Public License as published by
//    the Free Software Foundation; either version 2 of the License, or
//    (at your option) any later version.
//
//    CPPC is distributed in the hope that it will be useful,
//    but WITHOUT ANY WARRANTY; without even the implied warranty of
//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//    GNU General Public License for more details.
//
//    You should have received a copy of the GNU General Public License
//    along with CPPC; if not, write to the Free Software
//    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA



#include <util/configuration/configuration_manager.h>
#include <util/xml/xml_parser.h>
#include <util/xml/xml_parser_factory.h>

#include <cstdlib>

#include <iostream>

using std::cout;

using cppc::util::xml::XmlParserFactory;

namespace cppc {
  namespace util {
    namespace configuration {

      const ParameterKeyType ConfigurationManager::CPPC_PARAMETER_PREFIX(
        "CPPC" );
      string ConfigurationManager::CPPC_CONFIGURATION_FILE_PATH( "" );


      ConfigurationManager::ConfigurationManager()
        : configuration( ParameterMap() ), initialized( false ),
        parser( 0 ) {}

      ConfigurationManager::ConfigurationManager(
        const ConfigurationManager & cm )
        : configuration( cm.configuration ), initialized( cm.initialized ),
        parser( cm.parser ) {}

      ConfigurationManager & ConfigurationManager::operator=(
        const ConfigurationManager & rhs ) {

        if( this == &rhs ) {
          return *this;
        }

        configuration = rhs.configuration;
        initialized = rhs.initialized;
        parser = rhs.parser;

        return *this;
      }


      void ConfigurationManager::init( const ParameterVector & parameters ) {

        if( initialized ) {
          return;
        }

        ParameterIterator i = parameters.begin();

        while( i != parameters.end() ) {

          ParameterType parameter = (*i);
          configuration.insert( ParameterMap::value_type( parameter.key,
            parameter.value ) );

          i++;
        }

        // Now we have our command line parameters we look for the configuration
        // file name
        initCppcConfigurationFilePath();

        // Getting the XML Parser...
        parser = XmlParserFactory::instance().getParser(
          getParserTypeParameter() );
        parser->parseFile( CPPC_CONFIGURATION_FILE_PATH );

        initialized = true;
      }

      const ParameterValueType * const ConfigurationManager::getParameter(
        const ParameterKeyType & key ) {

        // 1st: Look in the cache
        const ParameterValueType * newParam = getParameterFromCache( key );

        // The parameter is already cached, so we can safely return
        if( newParam != 0 ) {
          return newParam;
        }

        // If it is not cached then we must find it and $ it

        // 2nd: Look in environment
        if( newParam == 0 ) {
          newParam = getParameterFromEnvironment( key );
        }

        // 3rd: Look in the configuration file
        if( newParam == 0 ) {
          newParam = getParameterFromConfigurationFile( key );
        }

        if( newParam != 0 ) {
          configuration.insert( ParameterMap::value_type( key, *newParam ) );
          return newParam;
        }

        // Maybe this politic could be changed
        cout << "Missing Parameter: " << key << "\n";
        exit( 0 );

        return NULL;

      }

      const ParameterValueType * ConfigurationManager::getParameterFromCache(
        const ParameterKeyType & key ) const {

        ParameterMap::const_iterator i = configuration.find( key );

        if( i != configuration.end() ) {
          return &(i->second);
        } else {
          return NULL;
        }
      }

      ParameterValueType * ConfigurationManager::getParameterFromEnvironment(
        const ParameterKeyType & key ) const {

        ParameterKeyType env_key( key );

        for( unsigned int i = 0; i < env_key.length(); i++ ) {
          if( env_key[i] == '/' ) {
            env_key[i] = '_';
          }
        }

        char * value = std::getenv( env_key.c_str() );

        if( value != NULL ) {
          return new ParameterValueType( value );
        }

        return NULL;

      }

      ParameterValueType *
        ConfigurationManager::getParameterFromConfigurationFile(
        const ParameterKeyType & key ) const {

        if( !initialized ) {
          return NULL;
        }

        string value = parser->getParameter( key );
        if( value == "" ) {
          return NULL;
        } else {
          return new string( value );
        }
      }

      void ConfigurationManager::writeStateToConfigurationFile() const {
      }

      const ParameterKeyType
        ConfigurationManager::CPPC_CONFIGURATION_FILE_PARAMETER_KEY(
          "CPPC/Configuration/FilePath" );
      void ConfigurationManager::initCppcConfigurationFilePath() {

        const ParameterValueType * cppcConfFile = getParameterFromCache(
          CPPC_CONFIGURATION_FILE_PARAMETER_KEY );
        if( cppcConfFile == NULL ) {
          cppcConfFile = getParameterFromEnvironment(
            CPPC_CONFIGURATION_FILE_PARAMETER_KEY );
        }
        if( cppcConfFile != NULL ) {
          CPPC_CONFIGURATION_FILE_PATH = *(cppcConfFile);
        } else {
          string defaultName( ".cppc_config" );
          ParameterValueType * homeDir = getParameterFromEnvironment( "HOME" );
          CPPC_CONFIGURATION_FILE_PATH = *homeDir + "/" + defaultName;
          cout << "CPPC: Trying to use default file " <<
            CPPC_CONFIGURATION_FILE_PATH << " for configuration. Override " <<
              "with CPPC/Configuration/FilePath\n";
          delete homeDir;
        }
      }

      const ParameterKeyType
        ConfigurationManager::CPPC_PARSER_TYPE_PARAMETER_KEY(
        "CPPC/XmlParser/Type" );
      ParserType ConfigurationManager::getParserTypeParameter() {

        const ParameterValueType * parserType = getParameterFromCache(
          CPPC_PARSER_TYPE_PARAMETER_KEY );
        if( parserType == NULL ) {
          parserType = getParameterFromEnvironment(
            CPPC_PARSER_TYPE_PARAMETER_KEY );
        }
        if( parserType != NULL ) {
          return atoi( parserType->c_str() );
        } else {
          return 0;
        }
      }

    }
  }
}
